Sources/StreamChat/Repositories/OfflineRequestsRepository.swift

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+typealias QueueOfflineRequestBlock = (DataEndpoint) -> Void
+typealias DataEndpoint = Endpoint<Data>
+
+extension Endpoint {
+    var withDataResponse: DataEndpoint {
+        DataEndpoint(
+            path: path,
+            method: method,
+            queryItems: queryItems,
+            requiresConnectionId: requiresConnectionId,
+            requiresToken: requiresToken,
+            body: body
+        )
+    }
+}
+
+/// OfflineRequestsRepository handles both the enqueuing and the execution of offline requests when needed.
+/// When running the queued requests, it basically passes the requests on to the APIClient, and waits for its result.
+class OfflineRequestsRepository {
+    let database: DatabaseContainer
+    let apiClient: APIClient
+
+    /// Serial queue used to enqueue pending requests one after another
+    private let retryQueue = DispatchQueue(label: "com.stream.queue-requests")
+
+    init(database: DatabaseContainer, apiClient: APIClient) {
+        self.database = database
+        self.apiClient = apiClient
+    }
+
+    /// - If the requests succeeds -> The request is removed from the pending ones
+    /// - If the request fails with a connection error -> The request is kept to be executed once the connection is back (we are not putting it back at the queue to make sure we respect the order)
+    /// - If the request fails with any other error -> We are dismissing the request, and removing it from the queue
+    func runQueuedRequests(completion: @escaping () -> Void) {
```

Comment:
This part is a bit confusing. If the goal is to not execute any API requests while the recovery flow, why it's not enough to just add them to `operationQueue` which gets suspended when we enter recovery flow and resumed when recovery is done?

Reply:
Discussed offline